modules/cli/src/main/scala/scala/cli/commands/Update.scala

```diff
@@ -50,48 +57,97 @@ object Update extends ScalaCommand[UpdateOptions] {
+    */
+  lazy val pathDuplicatesWarning: Either[String, String] = either {
+    s"""These are the instances of Scala CLI on your PATH:
+       |${value(pathInstances).mkString("\n")}
+       |If you have more than one instance, it is recommended to only keep the most recent version.""".stripMargin
+  }
+
+  def update(options: UpdateOptions, maybeScalaCliBinPath: Option[os.Path]): Unit = {
+    val maybeCurrentVersion = maybeScalaCliBinPath.map {
+      scalaCliBinPath =>
+        val res = os.proc(scalaCliBinPath, "version").call(cwd = os.pwd, check = false)
+        if (res.exitCode == 0)
+          res.out.text().trim
+        else
+          "0.0.0"
     }
 
-    lazy val newestScalaCliVersion = {
-      val resp = ProcUtil.downloadFile("https://github.com/VirtusLab/scala-cli/releases/latest")
+    val currentVersion = maybeCurrentVersion.getOrElse(Constants.version)
 
-      val scalaCliVersionRegex = "tag/v(.*?)\"".r
-      scalaCliVersionRegex.findFirstMatchIn(resp).map(_.group(1))
-    }.getOrElse(
-      sys.error("Can not resolve ScalaCLI version to update")
-    )
+    val maybeOutDated =
+      maybeCurrentVersion.map(CommandUtils.isOutOfDateVersion(newestScalaCliVersion, _))
 
-    val isOutdated = CommandUtils.isOutOfDateVersion(newestScalaCliVersion, currentVersion)
+    val outDated = maybeOutDated.getOrElse(isOutdated)
 
     if (!options.isInternalRun)
-      if (isOutdated)
+      if (outDated)
         updateScalaCli(options, newestScalaCliVersion)
       else println("ScalaCLI is up-to-date")
-    else if (isOutdated)
+    else if (outDated)
       println(
         s"""Your ScalaCLI $currentVersion is outdated, please update ScalaCLI to $newestScalaCliVersion
            |Run 'curl -sSLf https://virtuslab.github.io/scala-cli-packages/scala-setup.sh | sh' to update ScalaCLI.""".stripMargin
       )
   }
 
-  def checkUpdate(options: UpdateOptions) = {
-
-    val scalaCliBinPath = options.installDirPath / options.binaryName
-
-    lazy val execScalaCliPath = os.proc("which", "scala-cli").call(
+  /** Either an error or a Seq of the descriptions of scala-cli's PATH instances in linux
+    */
+  lazy val pathInstances: Either[String, List[String]] = {
+    val scalaCliPathInstances = os.proc("type", "-ap", "scala-cli").call(
       cwd = os.pwd,
       mergeErrIntoOut = true,
       check = false
-    ).out.text().trim
-    lazy val isScalaCliInPath = // if binDir is non empty, we not except scala-cli in PATH, it is useful in tests
-      execScalaCliPath.contains(options.installDirPath.toString()) || options.binDir.isDefined
+    )
+
+    if (scalaCliPathInstances.exitCode == 0)
+      Right(scalaCliPathInstances.out.text().trim.split(sys.props("line.separator")).toList)
+    else Left(scalaCliPathInstances.err.text().trim())
+  }
+
+  def checkUpdate(options: UpdateOptions): Unit = {
+
+    val scalaCliBinPath = options.installDirPath / options.binaryName
+
+    lazy val absolutePathToScalaCli: String = {
```

Comment:
also, why this is a lazy val inside def?

Reply:
yes. it would get calculated any time the def is called anyway. So perhaps, after the extraction to another different location, it would make sense to keep it lazy.

Reply:
solved!